lib/redis/connection/ruby.rb

```diff
@@ -49,39 +49,43 @@ def gets
       end
 
       def _read_from_socket(nbytes)
-        begin
-          read_nonblock(nbytes)
-        rescue IO::WaitReadable
-          if IO.select([self], nil, nil, @timeout)
-            retry
-          else
-            raise Redis::TimeoutError
-          end
-        rescue IO::WaitWritable
-          if IO.select(nil, [self], nil, @timeout)
-            retry
-          else
-            raise Redis::TimeoutError
+        loop do
+          case chunk = read_nonblock(nbytes, exception: false)
+          when :wait_readable
+            unless IO.select([self], nil, nil, @timeout)
+              raise Redis::TimeoutError
+            end
+          when :wait_writable
+            unless IO.select(nil, [self], nil, @timeout)
+              raise Redis::TimeoutError
+            end
+          when String
+            return chunk
           end
         end
       rescue EOFError
```

Comment:
https://docs.ruby-lang.org/en/2.7.0/IO.html#method-i-read_nonblock

> At EOF, it will return nil instead of raising EOFError.

Should we handle `nil` at `case` statement?